device/transport/mqtt/devdoc/mqtt_base_requirements.md

```diff
@@ -45,40 +46,61 @@ The `connect` method establishes a connection with the server using the config o
 
-### Mqtt.publish(message)
-The `publish` method publishes the message passed as argument.
+### MqttBase.unsubscribe(topic, callback)
 
-**SRS_NODE_COMMON_MQTT_BASE_12_006: [** The `publish` method shall throw `ReferenceError` “Invalid message” if the message is falsy. **]**
+**SRS_NODE_COMMON_MQTT_BASE_16_031: [** The `unsubscribe` method shall throw a `ReferenceError` if the `topic` argument is falsy. **]**
 
-**SRS_NODE_COMMON_MQTT_BASE_12_007: [** The `publish` method shall call `publish`  on MQTT.JS  library with the given message. **]**
+**SRS_NODE_COMMON_MQTT_BASE_16_027: [** The `unsubscribe` method shall call the callback with a `NotConnectedError` if the connection hasn't been established prior to calling `publish`. **]**
 
-**SRS_NODE_COMMON_MQTT_BASE_16_008: [** The `publish` method shall use a topic formatted using the following convention: `devices/<deviceId>/messages/events/`. **]**
+**SRS_NODE_COMMON_MQTT_BASE_16_028: [** The `unsubscribe` method shall call `unsubscribe` on the mqtt library and pass it the `topic`. **]**
 
-**SRS_NODE_COMMON_MQTT_BASE_16_009: [** If the message has properties, the property keys and values shall be uri-encoded, then serialized and appended at the end of the topic with the following convention: `<key>=<value>&<key2>=<value2>&<key3>=<value3>(...)`. **]**
+**SRS_NODE_COMMON_MQTT_BASE_16_029: [** The `unsubscribe` method shall call the `callback` argument with no arguments if the operation succeeds. **]**
 
-**SRS_NODE_COMMON_MQTT_BASE_16_010: [** The `publish` method shall use QoS level of 1. **]**
+**SRS_NODE_COMMON_MQTT_BASE_16_030: [** The `unsubscribe` method shall call the `callback` argument with an `Error` if the operation fails. **]**
 
-**SRS_NODE_COMMON_MQTT_BASE_16_011: [** The `publish` method shall serialize the `messageId` property of the message as a key-value pair on the topic with the key `$.mid`. **]**
 
-**SRS_NODE_COMMON_MQTT_BASE_16_012: [** The `publish` method shall serialize the `correlationId` property of the message as a key-value pair on the topic with the key `$.cid`. **]**
+### MqttBase.updateSharedAccessSignature(sharedAccessSignature, callback)
 
-**SRS_NODE_COMMON_MQTT_BASE_16_013: [** The `publish` method shall serialize the `userId` property of the message as a key-value pair on the topic with the key `$.uid`. **]**
+**SRS_NODE_COMMON_MQTT_BASE_16_032: [** The `updateSharedAccessSignature` method shall throw a `ReferenceError` if the `sharedAccessSignature` argument is falsy. **]**
 
-**SRS_NODE_COMMON_MQTT_BASE_16_014: [** The `publish` method shall serialize the `to` property of the message as a key-value pair on the topic with the key `$.to`. **]**
+**SRS_NODE_COMMON_MQTT_BASE_16_033: [** The `updateSharedAccessSignature` method shall disconnect and reconnect the mqtt client with the new `sharedAccessSignature`. **]**
```

Comment:
>disconnect and reconnect [](start = 87, length = 24)

are you sure disconnect->reconnect belongs at this level?  None of the other operations at this level have any connection side effects.

Reply:
I keep going back and forth on this one. elements to think about:
- in AMQP this is taken care of in the putToken call which is also on the base layer
- this also feels like it would be useful in the DPS transport, so it makes sense for it to be shared
- we're effectively replacing the mqtt client object hiding it below the MqttBase class but this is not supposed to trigger anything on the Mqtt object itself: unlike with AMQP there are no links to reestablish. since we set cleanSession to false, all subscribtions will be automatically reenabled.
- the only reason other methods don't have an effect on the connection is because we can't be sure to have a shared access signature when they are called. that case is specific since by definition this is what it's for. otherwise i'd happily handle auto-connect on publish in here too.

if you're thinking of a particular case where this would be detrimental please do let me know.

---
In reply to: [141150498](https://github.com/Azure/azure-iot-sdk-node/pull/90#discussion_r141150498) [](ancestors = 141150498)